optuna/visualization/_contour.py

```diff
@@ -303,4 +233,160 @@ def _generate_contour_subplot(
+        sorted_params = []
+    elif params is None:
+        sorted_params = sorted(all_params)
+    elif len(params) <= 1:
+        _logger.warning("The length of params must be greater than 1.")
+        sorted_params = list(params)
+    else:
+        for input_p_name in params:
+            if input_p_name not in all_params:
+                raise ValueError("Parameter {} does not exist in your study.".format(input_p_name))
+        sorted_params = sorted(set(params))
+
+    sub_plot_infos: List[List[_SubContourInfo]]
+    if len(sorted_params) == 2:
+        x_param = sorted_params[0]
+        y_param = sorted_params[1]
+        sub_plot_info = _generate_contour_subplot_info(trials, x_param, y_param, target)
+        sub_plot_infos = [[sub_plot_info]]
+    else:
+        sub_plot_infos = []
+        for i, y_param in enumerate(sorted_params):
+            sub_plot_infos.append([])
+            for x_param in sorted_params:
+                sub_plot_info = _generate_contour_subplot_info(trials, x_param, y_param, target)
+                sub_plot_infos[i].append(sub_plot_info)
+
+    return _ContourInfo(sorted_params=sorted_params, sub_plot_infos=sub_plot_infos)
+
+
+def _generate_contour_subplot_info(
+    trials: List[FrozenTrial],
+    x_param: str,
+    y_param: str,
+    target: Optional[Callable[[FrozenTrial], float]],
+) -> _SubContourInfo:
+
+    xaxis = _generate_axis_info(trials, x_param)
+    yaxis = _generate_axis_info(trials, y_param)
+
+    if x_param == y_param:
+        return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values=[[]])
+
+    if len(xaxis.indices) < 2:
+        _logger.warning("Param {} unique value length is less than 2.".format(x_param))
+        return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values=[[]])
+    if len(yaxis.indices) < 2:
+        _logger.warning("Param {} unique value length is less than 2.".format(y_param))
+        return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values=[[]])
+
+    z_values = [
+        [float("nan") for _ in range(len(xaxis.indices))] for _ in range(len(yaxis.indices))
+    ]
+
+    for i, trial in enumerate(trials):
+        if x_param not in trial.params or y_param not in trial.params:
+            continue
+        x_value = xaxis.values[i]
+        y_value = yaxis.values[i]
+        assert x_value is not None
+        assert y_value is not None
+        x_i = xaxis.indices.index(x_value)
+        y_i = yaxis.indices.index(y_value)
+
+        if target is None:
+            value = trial.value
+        else:
+            value = target(trial)
+
+        if isinstance(value, int):
+            value = float(value)
+        elif not isinstance(value, float):
+            raise ValueError(
+                f"Trial{trial.number} has COMPLETE state, but its target value is non-numeric."
+            )
+        z_values[y_i][x_i] = value
+
+    return _SubContourInfo(xaxis=xaxis, yaxis=yaxis, z_values=z_values)
```

Comment:
Currently, `_SubContourInfo` has the information of `z_values` as `List[List[float]]`, where nonexistent values are represented by `nan`. I think it would be more readable if it just had `z_map: Dict[Tuple[float, float], float]` instead of `z_value`.

Reply:
Thanks for the comment. It makes sense. I followed your suggestion. PTAL.